app/src/main/java/zapsolutions/zap/util/NfcUtil.java

```diff
@@ -0,0 +1,101 @@
+package zapsolutions.zap.util;
+
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.nfc.NdefMessage;
+import android.nfc.NdefRecord;
+import android.nfc.NfcAdapter;
+import android.nfc.tech.Ndef;
+import android.os.Parcelable;
+import android.widget.Toast;
+
+import java.util.Arrays;
+
+import zapsolutions.zap.R;
+
+public class NfcUtil {
+
+    private static final String LOG_TAG = NfcUtil.class.getName();
+
+    /**
+     * This function reads the content of an NFC Ndef Tag.
+     * As we want to read all NFC Tags while zap is open and a specific Activities (done with ForegroundDispatch),
+     * we have to make sure that we validate the tag type properly in this function.
+     * The filters from the manifest only apply to a tag that is read while the app is not in foreground
+     * or at an activity that does not do any foreground dispatching.
+     *
+     * @param ctx      Context.
+     * @param intent   The Intent
+     * @param listener listener which provides the payload in its onSuccess function
+     */
+    public static void readTag(Context ctx, Intent intent, OnNfcResponseListener listener) {
+        String action = intent.getAction();
+        ZapLog.debug(LOG_TAG, "onNewIntent: " + action);
+        if (action != null) {
+            if (action.equals(NfcAdapter.ACTION_TAG_DISCOVERED) || action.equals(NfcAdapter.ACTION_NDEF_DISCOVERED)) {
+
+                Parcelable[] rawMessages =
+                        intent.getParcelableArrayExtra(NfcAdapter.EXTRA_NDEF_MESSAGES);
+                if (rawMessages != null) {
+                    NdefMessage[] messages = new NdefMessage[rawMessages.length];
+                    for (int i = 0; i < rawMessages.length; i++) {
+                        messages[i] = (NdefMessage) rawMessages[i];
+                    }
+                    // Process the messages array.
+                    NdefMessage message = messages[0];
+
+                    ZapLog.debug(LOG_TAG, "Ndef message: " + message);
+                    NdefRecord[] records = message.getRecords();
+                    ZapLog.debug(LOG_TAG, "Ndef record: " + records[0]);
+                    if (records[0].getTnf() == NdefRecord.TNF_WELL_KNOWN) {
+                        if (Arrays.equals(records[0].getType(), NdefRecord.RTD_URI)) {
+                            byte[] rawPayload = records[0].getPayload();
+                            StringBuilder sb = new StringBuilder();
+                            for (int i = 1; i < rawPayload.length; i++) {
+                                sb.append((char) rawPayload[i]);
+                            }
+                            String payload = sb.toString();
+                            ZapLog.debug(LOG_TAG, "Ndef payload: " + payload);
+                            if (PrefsUtil.isWalletSetup()) {
```

Comment:
I would consider to move this check outside of the util.
It may lead to a little more code on the caller side, but a NFC util should not really care about the user having set up the wallet.

Reply:
:+1: